pageserver/src/layered_repository.rs

```diff
@@ -265,50 +282,97 @@ impl Repository for LayeredRepository {
 
         let timelines = self.timelines.lock().unwrap();
         for (timelineid, timeline) in timelines.iter() {
-            shutdown_timeline(*timelineid, timeline.as_ref())?;
+            shutdown_timeline(*timelineid, timeline)?;
         }
 
         Ok(())
     }
 
-    fn unload_timeline(&self, timeline_id: ZTimelineId) -> Result<()> {
+    fn set_timeline_state(&self, timeline_id: ZTimelineId, new_state: TimelineState) -> Result<()> {
         let mut timelines = self.timelines.lock().unwrap();
-        let removed_timeline = match timelines.remove(&timeline_id) {
-            Some(timeline) => timeline,
-            None => {
-                warn!("Timeline {} not found, nothing to remove", timeline_id);
-                return Ok(());
+        let old_timeline = timelines.remove(&timeline_id);
+        let mut timeline_to_shutdown = None;
+        match new_state {
+            TimelineState::Ready => {
+                let locked_timeline = self.get_timeline_locked(timeline_id, &mut timelines)?;
```

Comment:
It looks confusing that we remove timeline from timelines and then call get_timeline_locked which will look for this timeline in the map we just removed it from. Am I correct?

Reply:
It is indeed, I've rewritten it.